### PR TITLE
libpulseaudio: add module-aaudio-sink

### PR DESCRIPTION
**Pull request by tomty89**
Add a sink module that makes use of the AAudio API, which is
available in Oreo or later.

The sink will have its sample rate and latency configured based on
the info we got with the API. They will also be updated accordingly
when the output device is changed.

### CODE REVIEW

**Comment by its-pointless** [comment]
Fails because libaaudio.so isn't available for android 5 toolchain. Easily fixed by copying over aaudio lib at tool chain creation. 

Do you take requests?
Microphone support would be nice

**Comment by tomty89** [comment]
@its-pointless Might take a look in the future. But to be honest I am not particularly interested in implementing audio capturing, so most likely I would leave it to someone else. It would be a separate module anyway.

File: packages/libpulseaudio/build.sh [hunk 1]
```diff
@@ -17,12 +17,18 @@ TERMUX_PKG_EXTRA_CONFIGURE_ARGS="--disable-neon-opt
...
 TERMUX_PKG_CONFFILES="etc/pulse/client.conf etc/pulse/daemon.conf etc/pulse/default.pa etc/pulse/system.pa"
+TERMUX_PKG_API_LEVEL=26
 
 termux_step_pre_configure () {
 	mkdir $TERMUX_PKG_SRCDIR/src/modules/sles
 	cp $TERMUX_PKG_BUILDER_DIR/module-sles-sink.c $TERMUX_PKG_SRCDIR/src/modules/sles
+	mkdir $TERMUX_PKG_SRCDIR/src/modules/aaudio
+	cp $TERMUX_PKG_BUILDER_DIR/module-aaudio-sink.c $TERMUX_PKG_SRCDIR/src/modules/aaudio
+	sed -i "s/define __ANDROID_API__ $TERMUX_PKG_API_LEVEL/define __ANDROID_API__ 21/" \
+		$TERMUX_STANDALONE_TOOLCHAIN/sysroot/usr/include/android/api-level.h
```
[the remark this answers is not in the document]
Comment (tomty89) [reply packages/libpulseaudio/build.sh]: Hmm okay. Let me try build with that but without `ac_cv_func_posix_madvise`, and see if it fails as expected.

[the remark this answers is not in the document]
Comment (tomty89) [reply packages/libpulseaudio/build.sh]: Yup works as expected. Updated the commit.

File: packages/libpulseaudio/module-aaudio-sink.c [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+#include <pulsecore/core-util.h>
+#include <pulsecore/modargs.h>
+#include <pulsecore/log.h>
+#include <pulsecore/thread.h>
+#include <pulsecore/thread-mq.h>
+#include <pulsecore/rtpoll.h>
+
+#include <aaudio/AAudio.h>
+
+PA_MODULE_AUTHOR("Lennart Poettering, Nathan Martynov, Tom Yan");
```
[the remark this answers is not in the document]
Comment (tomty89) [reply packages/libpulseaudio/module-aaudio-sink.c]: Now both sink modules have only my name listed in `PA_MODULE_AUTHOR`. I suppose if anyone who thinks he deserves to have his name in there, he can always add it back with a PR.

[the remark this answers is not in the document]
Comment (twaik) [reply packages/libpulseaudio/module-aaudio-sink.c]: I think my name (char *name = "Nathan Martynov";) should be listed in SLES backend source...

[the remark this answers is not in the document]
Comment (tomty89) [reply packages/libpulseaudio/module-aaudio-sink.c]: Sorry what do you mean? So should I add `Nathan Martynov` back in `module-sles-sink`? (Aren't you `Twaik Yont` btw?)

**Comment by fornwall** [comment]
Thanks! The updated package is now available for installation.

**Comment by edward-p** [comment]
@tomty89 

`buffer underrun issue` with libao

```
...
I: [pulseaudio] sink-input.c: Created input 0 "libao[pianobar] playback stream" on AAudio_sink with sample spec s16le 2ch 44100Hz and channel map front-left,front-right
I: [pulseaudio] sink-input.c:     media.name = "libao[pianobar] playback stream"
I: [pulseaudio] sink-input.c:     application.name = "libao[pianobar]"
I: [pulseaudio] sink-input.c:     native-protocol.peer = "UNIX socket client"
I: [pulseaudio] sink-input.c:     native-protocol.version = "32"
I: [pulseaudio] sink-input.c:     application.process.id = "2352"
I: [pulseaudio] sink-input.c:     application.process.user = "u0_a114"
I: [pulseaudio] sink-input.c:     application.process.host = "localhost"
I: [pulseaudio] sink-input.c:     application.process.binary = "pianobar"
I: [pulseaudio] sink-input.c:     application.process.machine_id = "localhost"
I: [pulseaudio] sink-input.c:     module-stream-restore.id = "sink-input-by-application-name:libao[pianobar]"
I: [pulseaudio] protocol-native.c: Requested tlength=50.00 ms, minreq=12.49 ms
D: [pulseaudio] protocol-native.c: Adjust latency mode enabled, configuring sink latency to half of overall latency.
D: [pulseaudio] protocol-native.c: Requested latency=12.51 ms, Received latency=30.48 ms
D: [pulseaudio] memblockq.c: memblockq requested: maxlength=4194304, tlength=9784, base=4, prebuf=7584, minreq=2204 maxrewind=0
D: [pulseaudio] memblockq.c: memblockq sanitized: maxlength=4194304, tlength=9784, base=4, prebuf=7584, minreq=2204 maxrewind=0
I: [pulseaudio] protocol-native.c: Final latency 85.94 ms = 30.48 ms + 2*12.49 ms + 30.48 ms
D: [pulseaudio] sink.c: AAudio_sink: state: IDLE -> RUNNING
D: [pulseaudio] core-subscribe.c: Dropped redundant event due to change event.
D: [aaudio-sink] protocol-native.c: Requesting rewind due to end of underrun.
D: [AudioTrack] protocol-native.c: Implicit underrun of 'libao[pianobar] playback stream'
D: [aaudio-sink] protocol-native.c: Requesting rewind due to end of underrun.
D: [AudioTrack] protocol-native.c: Implicit underrun of 'libao[pianobar] playback stream'
D: [aaudio-sink] protocol-native.c: Requesting rewind due to end of underrun.
...
```

Seems there is something wrong with latency again? 

However, since applyed the patch [https://github.com/xiph/libao/pull/8](https://github.com/xiph/libao/pull/8) libao works fine with `module-seles-sink`

**Comment by edward-p** [comment]
OK, this can be fixed manually by set `buffer_time=280` to `libao.conf`

**Comment by tomty89** [comment]
I assume you are using Bluetooth audio?

I've seen device/ROM reports fast flag and a low buffer size for Bluetooth audio, and I doubt that they are valid.

**Comment by edward-p** [comment]
Yeah，but I just find my buffer size for none Bluetooth audio is also very low.

```
I: [pulseaudio] sink-input.c:     module-stream-restore.id = "sink-input-by-application-name:libao[pianobar]"
I: [pulseaudio] protocol-native.c: Requested tlength=50.00 ms, minreq=12.49 ms
D: [pulseaudio] protocol-native.c: Adjust latency mode enabled, configuring sink latency to half of overall latency.
D: [pulseaudio] protocol-native.c: Requested latency=12.51 ms, Received latency=16.00 ms
D: [pulseaudio] memblockq.c: memblockq requested: maxlength=4194304, tlength=7232, base=4, prebuf=5032, minreq=2204 maxrewind=0
D: [pulseaudio] memblockq.c: memblockq sanitized: maxlength=4194304, tlength=7232, base=4, prebuf=5032, minreq=2204 maxrewind=0
I: [pulseaudio] protocol-native.c: Final latency 57.00 ms = 16.01 ms + 2*12.49 ms + 16.00 ms
D: [pulseaudio] sink.c: AAudio_sink: state: IDLE -> RUNNING
D: [pulseaudio] core-subscribe.c: Dropped redundant event due to change event.
D: [aaudio-sink] protocol-native.c: Requesting rewind due to end of underrun.
D: [AudioTrack] protocol-native.c: Implicit underrun of 'libao[pianobar] playback stream'
D: [aaudio-sink] protocol-native.c: Requesting rewind due to end of underrun.
D: [AudioTrack] protocol-native.c: Implicit underrun of 'libao[pianobar] playback stream'
D: [aaudio-sink] protocol-native.c: Requesting rewind due to end of underrun.
D: [AudioTrack] protocol-native.c: Implicit underrun of 'libao[pianobar] playback stream'
D: [aaudio-sink] protocol-native.c: Requesting rewind due to end of underrun.
E: [AudioTrack] memblockq.c: Assertion 'bq->n_blocks >= 1' failed at /home/builder/.termux-build/libpulseaudio/src/src/pulsecore/memblockq.c:183, function void drop_block(pa_memblockq *, struct list_item *)(). Aborting.
[1]    7761 abort      pulseaudio -vvv --daemonize=no
```

I will test it on my other devices later.

**Comment by tomty89** [comment]
With wired audio low latency should work fine (at least way more likely).

**Comment by tomty89** [comment]
It is normal to get low latency and fast flag with wired audio (including USB). Not sure if it's a problem of your device or libao/pianobar.

**Comment by tomty89** [comment]
Btw the assertion failure is probably a separate issue. I've seen it occur with both modules. It's pretty random and I don't know what exactly triggers it though. Could be some upstream bug as well.

**Comment by edward-p** [comment]
Both wired audio and Bluetooth audio works fine on my old device without set latency manually. I think it's a problem of my new device. It should be an issue of the ROM. However, I assume that the manufacturer can achieve low latency on some high-end Bluetooth headphones which I don't have :(

**Pull request merged**